### PR TITLE
Rails Forms

### PR DESCRIPTION
**Pull request by julian3493**
- We create a user model with username, email, and password fields.
- We add validations for each field in the user model.
- We create a form template.
- We create index, new, show, edit, create, update, and destroy actions in the user controller.
- We create a view for new, index, edit, and show actions.

### CODE REVIEW

File: app/views/users/_form.html.erb [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+    <div id="error_explanation">
+      <h2><%= pluralize(user.errors.count, "error") %> prohibited this user from being saved:</h2>
+
+      <ul>
+        <% user.errors.full_messages.each do |message| %>
+          <li><%= message %></li>
+        <% end %>
+      </ul>
+    </div>
+  <% end %>
```
Comment (acushlakoncept) [inline app/views/users/_form.html.erb]: View-related logic can be encapsulated in View helpers (app/helpers/*_helper.rb)
* [x] There should be absolutely NO logic in views. >Views are only for displaying information that became available to it from its corresponding controller not generating it. **Resources**
* [Where Do I Put My Code?](http://codefol.io/posts/Where-Do-I-Put-My-Code)
* [The Beginner's Guide to Rails Helpers](https://mixandgo.com/learn/the-beginners-guide-to-rails-helpers)
*[Rails View Helpers Best Practices  @Guide](https://medium.com/@acushla4real/rails-view-helpers-best-practices-guide-d7d7b6db3db4)
* [screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/section/no_logic_in_helpers_requirement.png) from Pathwright

**Review by acushlakoncept — CHANGES_REQUESTED** [review]
# Status: Changes Required ♻️

#### Nice work team 👏  

You've done an awesome job so far 👌. But there are still some changes that are necessary for this project to be approved.

- [x] Your route in config/routes.rb should contain:  resources :users, only: [:new, :create, :edit, :update] ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/forms/user_route1.png) [screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/forms/user_route2.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/forms))

#### It is important to know how building forms evolved to form_for 
([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/forms/user_form1.png) [screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/forms/user_form2.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/forms)) Let's add these missing piece of code to our `_form.html.erb` file and we will be done.

* A commented out form built without a rails helper using 
`users_path` helper for action ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/forms/user_form1.png) [screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/forms/user_form2.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/forms))

* A commented out form built with `form_tag` a rails helper ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/forms/user_form_tag2.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/forms))

* Finally a form with `form_for` tag using `@user` from the controller ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/forms/user_form_for.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/forms))

#### Kindly refer to the comment I made on your code base about views related logic

### Cheers and Happy coding! 💻 👏

(Feel free to contact me on slack(@acushla), if you have any additional question)
Uduak Essien
[Twitter](https://twitter.com/acushlakoncept) | [LinkedIn](https://www.linkedin.com/in/acushlakoncept/)

File: app/controllers/users_controller.rb [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+  def edit; end
+
+  def create
+    @user = User.new(user_params)
+
+    respond_to do |format|
+      if @user.save
+        format.html { redirect_to @user, notice: 'User was successfully created.' }
+        format.json { render :show, status: :created, location: @user }
```
Comment (ghost) [inline app/controllers/users_controller.rb:22]: It should `redirect_to new_user_path` if a new user is successfully saved.

File: app/controllers/users_controller.rb [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+        format.json { render json: @user.errors, status: :unprocessable_entity }
+      end
+    end
+  end
+
+  def update
+    respond_to do |format|
+      if @user.update(user_params)
+        format.html { redirect_to @user, notice: 'User was successfully updated.' }
+        format.json { render :show, status: :created, location: @user }
```
Comment (ghost) [inline app/controllers/users_controller.rb:35]: It should `redirect_to edit_user_path` if the user was updated successfully.

File: app/views/users/_form.html.erb [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+      <label for="password">Password</label>
+      <input type="password" id="password">
+    </div>
+
+    <button class="actions" type="submit">
+      sutmit
+    </button>
+  <form> 
+<%
+=end%>
```
Comment (ghost) [inline app/views/users/_form.html.erb]: This code should be commented out.

File: app/views/users/_form.html.erb [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+      <%= form.label (:password, "Password") %>
+      <%= password_field_tag (:password) %>
+    </div>
+
+    <div class="actions">
+      <%= submit_tag ("Submit") %>
+    </div>
+  <% end %> 
+<%
+=end%>
```
Comment (ghost) [inline app/views/users/_form.html.erb]: This code should be commented out.

**Review by ghost — CHANGES_REQUESTED** [review]
# Status: Changes Required

### Hello team 😃,

**Good job implementing the requested changes 💪. However, there are still a few fixes needed.**

1). **I get the following error when I visit `new_user_path` and `edit_user_path`.**
pdfError

2). **Add the following in your Readme getting started manual.**

- Add the command for running migrations.

- Show the routes one needs to visit in order to create and edit a user.

- [optional] You can explain or provide documentation that is useful for installing `yarn` and postrgres.

3) **Check more in the comments below.**

### Happy coding 😃.

**PS: You can reach me on slack (Rose Wanjohi).**

**Review by addod19 — CHANGES_REQUESTED** [review]
## Status: Require changes :recycle: :traffic_light: 

Hi @julian3493 and @kenderb, great job fixing most of the comments outlined bt the previous TSE's :smile: The form is working as expected, there is one thing I would like us to fix before proceeding: :+1:  Kindly check the comment below
- Let us use only the working controller methods, for example in our `users controller` we have defined the `show` and `edit` method but it is empty and not working so it is not advisable to declare it as a resource in our routes

Kindly submit another code review once this is fixed :smile: 
_Happy coding_ :man_technologist:
[Daniel Larbi Addo](https://www.linkedin.com/in/daniel-larbi-addo/)

**Review by addod19 — APPROVED** [review]
## Status: Approved :trophy: :balloon: 

Hi @julian3493 and @kenderb, congratulations on a great job done :fireworks: I hereby approve the project :slightly_smiling_face: 

Go ahead and merge into development :+1: 
_Happy coding_ :man_technologist: 
[Daniel Larbi Addo](linkedin.com/in/daniel-larbi-addo/)

**Pull request merged**